### ISSUE TITLE
queue contains only garbage: Not enough bytes left to unpack, closing connection when diam message comes in two frames

### ISSUE THREAD

**Issue by BungaaFACE** [mensonen/diameter#14]
Hello, my diam connection closes periodicaly.

Log shows the following:
- 2024-08-01T15:03:40.531846 | WARNING | __init__.py | 1887 | log | <PeerConnection_instance> queue contains only garbage: Not enough bytes left to unpack, closing connection  
  
And after 10 sec:  
- 2024-08-01T15:03:51.891669 | INFO | node.py | 831 | _reconnect_peers | connection to "_dsr_fqdn_" has been lost for 11 seconds, reconnecting
- 2024-08-01T15:03:51.892541 | INFO | node.py | 292 | _add_peer_connection | added a new connection ['_dsr_ip_']:_dsr_port_ for peer Peer<(_dsr_fqdn_)>

In tcpdump i see
![image](https://github.com/user-attachments/assets/9ba5fc28-cb44-4408-bf6c-f622afc7dacc)
frame 2909 - first part of incoming diameter message
frame 2910 - outgoing tcp ack
frame 2911 - second part of incoming diameter message
frame 2912 - outgoing tcp ack
frame 2913 - outgoing tcp rst

My guess is that socket read happened between two frames of diameter, so diameter message wasn't fully recieved and got exception Not enough bytes left to unpack.

There are also other splitted diameter messages that handled succesfully, so i think it's just about timing
I cant send trace, but if you have any questions or suggestions i'm here. Please help to fix this issue

PS diameter message is ok

**Comment by mensonen** [comment]
Having the message be split across multiple frames shouldn't normally be an issue; the socket reads just dump bytes into an internal queue and diameter messages are constructed independently of this - and only after there are enough bytes in the queue to construct at least one message.

Can you share more log lines? There should be messages above the "garbage" warning line, mentioning "read X bytes",  "message incomplete" and/or "expecting a message with command code", perhaps even multiple times. These should pinpoint if the `Not enough bytes left to unpack` exception originates from parsing the message header or the entire message.

**Comment by BungaaFACE** [comment]
Here is debug logs:
- 2024-08-02T11:43:30.151338+0300 | DEBUG | __init__.py | 1887 | log | <PeerConnection() read 1012 bytes | {}
- 2024-08-02T11:43:30.151630+0300 | DEBUG | __init__.py | 1887 | log | <PeerConnection() expecting a message with command code 272, length 1020 | {}
- 2024-08-02T11:43:30.151712+0300 | DEBUG | __init__.py | 1887 | log | <PeerConnection() message incomplete (received 1012 bytes so far), waiting | {}
- 2024-08-02T11:43:30.151907+0300 | WARNING | __init__.py | 1887 | log | <PeerConnection() queue contains only garbage: Not enough bytes left to unpack, closing connection | {}
- 2024-08-02T11:43:30.152240+0300 | DEBUG | node.py | 1145 | remove_peer_connection | <PeerConnection() removed | {}

So resume_waiting works correctly and exception occurs when reading message.

```
msg_header = MessageHeader.from_bytes(self._read_buffer)
self.logger.debug(
    f"expecting a message with command code "
    f"{msg_header.command_code}, length {msg_header.length}")
if len(self._read_buffer) < msg_header.length:
    self.logger.debug(
        f"message incomplete (received "
        f"{len(self._read_buffer)} bytes so far), waiting")
    resume_waiting = True
message = Message.from_bytes(self._read_buffer[:msg_header.length])
self.reset_last_message()
self._read_buffer = self._read_buffer[msg_header.length:]
```

**Comment by mensonen** [comment]
Okay I think I spotted a rather obvious bug in there. I've fixed it but haven't yet merged my changes. Everything runs on my test machines same as before, but I can't reliably reproduce your issue to confirm that I've actually fixed what you reported. 

Can you try with the attached dev build and see if the problem is resolved for you as well? 

You should do `pip uninstall python-diameter` and then after unpacking the zip file `pip install python_diameter-0.7.3.dev0+gbd1d985.d20240802-py3-none-any.whl`.

[python_diameter-0.7.3.dev0+gbd1d985.d20240802.zip](https://github.com/user-attachments/files/16472662/python_diameter-0.7.3.dev0%2Bgbd1d985.d20240802.zip)

**Comment by BungaaFACE** [comment]
I cant install this version because its prod and we are using version 0.5.4. 
Purpose of old version is that we implemented add_realm function for Node to accept diam messages with other realm that peer has.

For newer version you changed a bit self._peer_routes, so this implementation dont work on new version.

Did you just added `else` before reading message and nothing more? i can change it manually.

Also it would be great if you add possibility to add served origin realms to the Node. For now we just adding `self._peer_routes[realm_name][app] = peers` as where app and peers parameters from standart peer. I can make another issue with this feat request

**Comment by mensonen** [comment]
Yeah it's just an `else` that was missing:

```diff
diff --git a/src/diameter/node/peer.py b/src/diameter/node/peer.py
index fc975a9..49d08d4 100644
--- a/src/diameter/node/peer.py
+++ b/src/diameter/node/peer.py
@@ -569,9 +569,11 @@ class PeerConnection:
                             f"message incomplete (received "
                             f"{len(self._read_buffer)} bytes so far), waiting")
                         resume_waiting = True
-                    message = Message.from_bytes(self._read_buffer[:msg_header.length])
-                    self.reset_last_message()
-                    self._read_buffer = self._read_buffer[msg_header.length:]
+                    else:
+                        message = Message.from_bytes(
+                            self._read_buffer[:msg_header.length])
+                        self.reset_last_message()
+                        self._read_buffer = self._read_buffer[msg_header.length:]
 
                 except Exception as e:
                     if msg_header and len(self._read_buffer) >= msg_header.length:
```

Let me know if this fixes it for you and I'll add it to main branch.

Regarding the multiple realms and your added changes, please do create another issue for that. 

The `0.7.1` release actually added quite a lot of changes for working with multiple realms simultaneously so perhaps your case is already covered. Also `0.5.4` is really old, it was an early pre-release.

**Comment by BungaaFACE** [comment]
Looks like this fixed the issue. How long will it take to make release with this fix?

**Comment by mensonen** [comment]
I'd like to wait for #15 to be closed, then I can include that in the release as well.

**Comment by mensonen** [comment]
Fixed in v0.7.3, released today.
